homeassistant/components/plugwise/usb.py

```diff
@@ -0,0 +1,179 @@
+"""Support for Plugwise devices connected to a Plugwise USB-stick."""
+import asyncio
+import logging
+
+import plugwise
+from plugwise.exceptions import (
+    CirclePlusError,
+    NetworkDown,
+    PortError,
+    StickInitError,
+    TimeoutException,
+)
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import ATTR_STATE, EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.entity import Entity
+
+from .const import (
+    AVAILABLE_SENSOR_ID,
+    CONF_USB_PATH,
+    DOMAIN,
+    PLATFORMS_USB,
+    PW_TYPE,
+    STICK,
+    UNDO_UPDATE_LISTENER,
+    USB,
+    USB_SENSORS,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry_usb(hass: HomeAssistant, config_entry: ConfigEntry):
+    """Establish connection with plugwise USB-stick."""
+    hass.data.setdefault(DOMAIN, {})
+
+    def discover_finished():
+        """Create entities for all discovered nodes."""
+        nodes = stick.nodes()
+        _LOGGER.debug(
+            "Successfully discovered %s out of %s registered nodes",
+            str(len(nodes)),
+            str(stick.registered_nodes()),
+        )
+        for component in PLATFORMS_USB:
+            hass.data[DOMAIN][config_entry.entry_id][component] = []
+            for mac in nodes:
+                if component in stick.node(mac).get_categories():
+                    hass.data[DOMAIN][config_entry.entry_id][component].append(mac)
+            hass.async_create_task(
+                hass.config_entries.async_forward_entry_setup(config_entry, component)
+            )
+        stick.auto_update()
+        # Enable reception of join request and automatically accept new node join requests
+        stick.allow_join_requests(True, True)
+
+    def shutdown(event):
+        hass.async_add_executor_job(stick.disconnect)
+
+    stick = plugwise.stick(config_entry.data[CONF_USB_PATH])
+    hass.data[DOMAIN][config_entry.entry_id] = {PW_TYPE: USB, STICK: stick}
+    try:
+        _LOGGER.debug("Connect to USB-Stick")
+        await hass.async_add_executor_job(stick.connect)
+        _LOGGER.debug("Initialize USB-stick")
+        await hass.async_add_executor_job(stick.initialize_stick)
+        _LOGGER.debug("Discover Circle+ node")
+        await hass.async_add_executor_job(stick.initialize_circle_plus)
```

Comment:
Please avoid multiple jumps into the executor. Wrap them into a single function.